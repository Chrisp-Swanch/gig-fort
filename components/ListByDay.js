import { useState, useMemo } from "react";
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity
} from "react-native";
import { useGigs } from "../hooks/useGigs";
import GigsByDay from "./GigsByDay";
import GigsByWeek from "./GigsByWeek";
import { format,addDays } from "date-fns";

const ListByDay = ({ navigation }) => {
  const [currentDateMs, setCurrentDateMs] = useState(Date.now());
  const [showWeek, setShowByWeek] = useState(false);
  const gigs  = useGigs();


  //Filtering through gigs to return only current day's gigs
  const gigsToday = gigs.filter((gig) => { 
    const formattedDate = format(new Date(currentDateMs), 'do MMMM Y')
    const formattedGigDate = format(new Date(gig.dateAndTime.seconds*1000) ,'do MMMM Y')
    return formattedGigDate === formattedDate
  })


  //Generating date a week from current date
  const weekFromNow = addDays(currentDateMs,7)


  //Filtering through gigs and returning gigs for week after current date
  const gigsThisWeek = gigs.filter((gig) => {
    return gig.dateAndTime.seconds*1000 < weekFromNow && gig.dateAndTime.seconds*1000 >= currentDateMs
  })


  const gigsThisWeek_sorted = gigsThisWeek.sort((a,b) => a.dateAndTime - b.dateAndTime)

  const gigsThisWeek_newDate = gigsThisWeek_sorted.map((item) => {
    const formattedDate = format(new Date(item.dateAndTime.seconds*1000),'EEE LLL do Y')
    return {...item, dateAndTime: formattedDate}
  })


  //group gigs by date
  const gigsThisWeek_grouped = gigsThisWeek_newDate.reduce((acc, curr) => {
    if (acc[curr.dateAndTime]) {
      acc[curr.dateAndTime].push(curr);
    } else {
      acc[curr.dateAndTime] = [curr];
    }
    return acc;
  }, {});


  //conditionally renders either gig list by day or list by week
  const gigsToRender = showWeek ? (
    <GigsByWeek gigsThisWeek_grouped={gigsThisWeek_grouped} navigation={navigation} />
  ) : (
    <GigsByDay navigation={navigation} gigsFromSelectedDate={gigsToday} />
  );
  

  return (
    <View>
      <View style={styles.buttonContainer}>
        <TouchableOpacity
          onPress={() => setShowByWeek(false)}
          style={styles.touchable}
        >
          <Text style={showWeek ? null : styles.selected}>Gigs today</Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={() => setShowByWeek(true)}
          style={styles.touchable}
        >
          <Text style={showWeek ? styles.selected : null}>Gigs this week</Text>
        </TouchableOpacity>
      </View>
      {gigsToRender}
    </View>
  );
};

const styles = StyleSheet.create({
  gigCard: {
    marginBottom: 5,
    padding: 12,
    width:'85%',
  },
  header: {
    padding: 10,
  },
  buttonContainer: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-around",
    padding: 15,
  },
  touchable: {
    padding: 6,
  },
  gigCard_header: {
    fontFamily: "Sofia-Pro",
    fontSize: 17,
  },
  gigCard_details: {
    fontFamily: "Helvetica-Neue",
    color: "#778899",
    flexShrink: 1,
  },
  date: {
    paddingLeft: 10,
    fontFamily: "Sofia-Pro",
    fontSize: 20,
    textDecorationLine: "underline",
  },
  gigCard_items: {
    flexDirection:'row',
    alignItems:'center',
  },
  gigCard_items_img:{
    height:30,
    width:30
  },
  selected: {
    backgroundColor: "#68912b",
    padding: 5,
    color: "white",
    fontFamily: "Helvetica-Neue",
    borderRadius: 5,
  }
});

export default ListByDay;
