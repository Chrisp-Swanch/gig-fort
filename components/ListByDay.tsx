import { FC } from "react";
import { useState, useMemo } from "react";
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity
} from "react-native";
import { useGigs } from "../hooks/useGigs";
import GigsByDay from "./GigsByDay";
import GigsByWeek from "./GigsByWeek";
import { format,addDays } from "date-fns";
import { listProps } from "../routes/homeStack";

type ListScreenNavigationProp = listProps['navigation']

interface Props {
  navigation:ListScreenNavigationProp
}

const ListByDay:FC<Props> = ({ navigation }):JSX.Element => {

  const [currentDateMs, setCurrentDateMs] = useState<number>(Date.now());
  const [showWeek, setShowByWeek] = useState<boolean>(false);
  const gigs  = useGigs();



  //Filtering through gigs to return only current day's gigs
  const gigsToday = gigs.filter((gig) => { 
    const formattedDate = format(new Date(currentDateMs), 'do MMMM Y')
    const formattedGigDate = format(new Date(gig.dateAndTime.seconds*1000) ,'do MMMM Y')
    return formattedGigDate === formattedDate
  })

  //Generating date a week from current date
  const weekFromNow = addDays(currentDateMs,7)


  //Filtering through gigs and returning gigs for week after current date
  const gigsThisWeek = gigs.filter((gig) => {
    const gigDate = new Date(gig.dateAndTime.seconds*1000)
    return gigDate < weekFromNow && gig.dateAndTime.seconds*1000 >= currentDateMs
  })


  const gigsThisWeek_sorted = gigsThisWeek.sort((a,b) => a.dateAndTime.seconds - b.dateAndTime.seconds)

  const gigsThisWeek_newDate = gigsThisWeek_sorted.map((item) => {
    const formattedDate = format(new Date(item.dateAndTime.seconds*1000),'EEE LLL do Y')
    return {...item, titleDate: formattedDate}
  })


  //group gigs by date
  const gigsThisWeek_grouped = gigsThisWeek_newDate.reduce((acc, curr) => {
    if (acc[curr.titleDate]) {
      acc[curr.titleDate].push(curr);
    } else {
      acc[curr.titleDate] = [curr];
    }
    return acc;
  }, {});


  //conditionally renders either gig list by day or list by week
  const gigsToRender = showWeek ? (
    <GigsByWeek gigsThisWeek_grouped={gigsThisWeek_grouped} navigation={navigation} />
  ) : (
    <GigsByDay navigation={navigation} gigsFromSelectedDate={gigsToday} />
  );
  

  return (
    <View>
      <View style={styles.buttonContainer}>
        <TouchableOpacity
          onPress={() => setShowByWeek(false)}
          style={styles.touchable}
        >
          <Text style={showWeek ? null : styles.selected}>Gigs today</Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={() => setShowByWeek(true)}
          style={styles.touchable}
        >
          <Text style={showWeek ? styles.selected : null}>Gigs this week</Text>
        </TouchableOpacity>
      </View>
      {gigsToRender}
    </View>
  );
};

const styles = StyleSheet.create({
  gigCard: {
    marginBottom: 5,
    padding: 12,
    width:'85%',
  },
  header: {
    padding: 10,
  },
  buttonContainer: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-around",
    padding: 15,
  },
  touchable: {
    padding: 6,
  },
  gigCard_header: {
    fontFamily: "Sofia-Pro",
    fontSize: 17,
  },
  gigCard_details: {
    fontFamily: "Helvetica-Neue",
    color: "#778899",
    flexShrink: 1,
  },
  date: {
    paddingLeft: 10,
    fontFamily: "Sofia-Pro",
    fontSize: 20,
    textDecorationLine: "underline",
  },
  gigCard_items: {
    flexDirection:'row',
    alignItems:'center',
  },
  gigCard_items_img:{
    height:30,
    width:30
  },
  selected: {
    backgroundColor: "#68912b",
    padding: 5,
    color: "white",
    fontFamily: "Helvetica-Neue",
    borderRadius: 5,
  }
});

export default ListByDay;
