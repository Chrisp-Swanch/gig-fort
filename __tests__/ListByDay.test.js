import ListByDay from "../components/ListByDay";
import { render } from '@testing-library/react-native'
import "@testing-library/jest-dom/extend-expect"
import { format } from 'date-fns'

test("should render listByDay", () => {
    const { getByTestId } = render(<ListByDay/>)
    const header = getByTestId('header')
    expect(header).toBeTruthy()
})



    test('should filter gigs by date', () => {
      // create an array of gigs
      const gigs = [
        { dateAndTime: { seconds: 1609459200 } }, // 1 September 2021
        { dateAndTime: { seconds: 1609545600 } }, // 2 September 2021
        { dateAndTime: { seconds: 1609632000 } }, // 3 September 2021
      ]
  
      // create a new date object with the current date
      const currentDate = new Date(2021, 0, 1) // 1 September 2021
  
      // pass the gigs and currentDate to the filter function
      const gigsToday = gigs.filter((gig) => { 
        const formattedDate = format(currentDate, 'do MMMM Y')
        const formattedGigDate = format(new Date(gig.dateAndTime.seconds*1000) ,'do MMMM Y')
        return formattedGigDate === formattedDate
      })
  
      // check that the filter function returned the correct gigs
      expect(gigsToday).toStrictEqual([{ dateAndTime: { seconds: 1609459200 } }])
    })
