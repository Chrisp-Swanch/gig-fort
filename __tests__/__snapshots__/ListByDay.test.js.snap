// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": 15,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 6,
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#68912b",
            "borderRadius": 5,
            "color": "white",
            "fontFamily": "Helvetica-Neue",
            "padding": 5,
          }
        }
      >
        Gigs today
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 6,
        }
      }
    >
      <Text
        style={null}
      >
        Gigs this week
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
